independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java

```diff
@@ -69,13 +62,22 @@ public void init() {
 
     @Override
     public byte[] getResourceData(String resource) {
```

Comment:
this can become a method on JarFileReference, actually, hiding the ownershipt logic within it. 
It allows reading a reasouce (with its own retain/release pairs, to be able to close the file, if the release grant rights to do it...)

In order to achieve it we need to pass the AtomicReference<JarFileReference> to this method so it can "optmistiacally" attempts a compareAndSet(this, null) to it. 
If it fails is not a problem, because it means another thread has succeeded to create a new one OR nulling it out